contracts/Governance.sol

```diff
@@ -0,0 +1,285 @@
+        lastTimeQuorumReached = block.timestamp;
+
+        proposalList = _arbitrablePermissionList;
+        pinakion = _pinakion;
+        tokenController = _tokenController;
+
+        proposalQuorum = _proposalQuorum;
+        currentProposalQuorum = proposalQuorum;
+
+        quorumDivideTime = _quorumDivideTime;
+        votingTime = _votingTime;
+    }
+
+
+    // ****************************** //
+    // *          Modifiers         * //
+    // ****************************** //
+
+    modifier onlyWhenProposalIsNew(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.New, "Only when proposal in New state.");
+        _;
+    }
+
+    modifier onlyWhenProposalPutToSupport(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToSupport, "Only when proposal in PutToSupport state.");
+        _;
+    }
+
+    modifier onlyWhenProposalPutToVote(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        _;
+    }
+
+    modifier onlyWhenProposalIsDecided(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.Decided, "Only when proposal in Decided state.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriod(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp <= proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period.");
+        _;
+    }
+
+    modifier onlyWhenPermitted(bytes32 _id) {
+        require(proposalList.isPermitted(_id), "Only when proposal is permitted.");
+        _;
+    }
+
+    modifier onlyWhenQuorumReached(bytes32 _id) {
+        require(proposals[_id].quorumToken.balanceOf(supportDeposit) >= getRequiredQuorum(_id), "Only when propsal has quorum.");
+        _;
+    }
+
+    modifier onlyWhenVotingPeriodPassed(bytes32 _id) {
+        require(proposals[_id].state == ProposalState.PutToVote, "Only when proposal in PutToVote state.");
+        require(block.timestamp > proposals[_id].whenPutToVote + currentVotingTime, "Only when voting period passed.");
+        _;
+    }
+
+    modifier onlyWhenApproved(bytes32 _id) {
+        require(proposals[_id].approved, "Only when proposal is approved.");
+        _;
+    }
+
+
+    // ****************************** //
+    // *           Events           * //
+    // ****************************** //
+
+    event ProposalPutToSupport(bytes32 indexed _id);
+
+    event ProposalPutToVote(bytes32 indexed _id);
+
+    event ProposalDecided(bytes32 indexed _id, bool _approved);
+
+    event ProposalExecuted(bytes32 indexed _id);
+
+    event ProposalCreated(bytes32 indexed _id, address _destination, uint _amount, bytes _data, string _uriDescription, bytes32 _descriptionHash, string _uriArguments, bytes32 _argumentsHash);
```

Comment:
It's better to define events in order, so this one should go first.

Most of the time, you won't need all those params. You'll just fetch the events and then request details for the ones you care about. So I think `indexed _id` and `indexed _destination` is sufficient.
They cover the two use cases:
- Find out when a proposal was created from `_id`.
- Find all proposals for a `_destination`.